DigitalLearningSolutions.Data/Services/RegistrationService.cs

```diff
@@ -313,35 +315,46 @@ private void CreateDelegateAccountForAdmin(AdminRegistrationModel registrationMo
-            var failureIfAny = DelegateCreationError.FromStoredProcedureErrorCode(candidateNumberOrErrorCode);
-            if (failureIfAny != null)
+            try
             {
-                logger.LogError(
-                    $"Delegate account could not be created (error code: {candidateNumberOrErrorCode}) with email address: {registrationModel.PrimaryEmail}"
+                var candidateNumber =
+                    registrationDataService.RegisterNewUserAndDelegateAccount(delegateRegistrationModel);
+                passwordDataService.SetPasswordByCandidateNumber(
+                    candidateNumber,
+                    delegateRegistrationModel.PasswordHash!
                 );
 
-                throw new DelegateCreationFailedException(failureIfAny);
-            }
-
-            passwordDataService.SetPasswordByCandidateNumber(
-                candidateNumberOrErrorCode,
-                delegateRegistrationModel.PasswordHash!
-            );
+                // We know this will give us a non-null user.
+                // If the delegate hadn't successfully been added we would have errored out of this method earlier.
+                var delegateUser = userDataService.GetDelegateUserByCandidateNumber(
+                    candidateNumber,
+                    delegateRegistrationModel.Centre
+                )!;
 
-            // We know this will give us a non-null user.
-            // If the delegate hadn't successfully been added we would have errored out of this method earlier.
-            var delegateUser = userDataService.GetDelegateUserByCandidateNumber(
-                candidateNumberOrErrorCode,
-                delegateRegistrationModel.Centre
-            )!;
+                userDataService.UpdateDelegateProfessionalRegistrationNumber(
+                    delegateUser.Id,
+                    registrationModel.ProfessionalRegistrationNumber,
+                    true
+                );
 
-            userDataService.UpdateDelegateProfessionalRegistrationNumber(
-                delegateUser.Id,
-                registrationModel.ProfessionalRegistrationNumber,
-                true
-            );
+                return userDataService.GetUserIdFromDelegateId(delegateUser.Id);
+            }
+            catch (DelegateCreationFailedException e)
+            {
+                var error = e.Error;
+                logger.LogError(
+                    $"Delegate account could not be created for admin with email address: {registrationModel.PrimaryEmail}. Failure: {error.Name}."
+                );
+                throw new DelegateCreationFailedException(error);
+            }
+            catch
+            {
+                var error = DelegateCreationError.UnexpectedError;
+                logger.LogError(
+                    $"Could not create delegate account for admin. Failure: {error.Name}."
+                );

```

Comment:
These exception logs should log the exception. There is an overload to do so. The throw new should also really be making use of the innerException overload.

Reply:
Done, let me know if that's what you were going for

Reply:
I think the messages we probably fine before, but the logging is correct now.